streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java

```diff
@@ -17,46 +17,96 @@
+     * exception}.
      *
-     * @param context the context; may not be null
+     * @param context the context
      */
     void init(final ProcessorContext context);
 
     /**
-     * Transform the record with the given key and value.
+     * Transform the given value to a new value.
+     * Additionally, any {@link org.apache.kafka.streams.processor.StateStore state} that is
+     * {@link KStream#transformValues(ValueTransformerSupplier, String...) attached} to this operator can be accessed
+     * and modified arbitrarily (cf. {@link ProcessorContext#getStateStore(String)}).
+     * <p>
+     * Note, that using {@link ProcessorContext#forward(Object, Object)},
+     * {@link ProcessorContext#forward(Object, Object, int)}, and
+     * {@link ProcessorContext#forward(Object, Object, String)} is not allowed within {@code transform} and
+     * will result in an {@link org.apache.kafka.streams.errors.StreamsException exception}.
      *
-     * @param value the value for the record
-     * @return new value
+     * @param value the value to be transformed
+     * @return the new value
      */
-    R transform(final V value);
+    VR transform(final V value);
 
     /**
-     * Perform any periodic operations and possibly return a new value, if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context
-     * during {@link #init(ProcessorContext) initialization}.
+     * Perform any periodic operations if this processor {@link ProcessorContext#schedule(long) schedule itself} with
+     * the context during {@link #init(ProcessorContext) initialization}.
+     * <p>
+     * It is not possible to return any new output records within {@code punctuate}.
+     * Using {@link ProcessorContext#forward(Object, Object)}, {@link ProcessorContext#forward(Object, Object, int)},
+     * or {@link ProcessorContext#forward(Object, Object, String)} will result in an
+     * {@link org.apache.kafka.streams.errors.StreamsException exception}.
+     * Furthermore, {@code punctuate} must return {@code null}.
+     * <p>
+     * Note, that {@code punctuate} is called base on <it>stream time</it> (i.e., time progress with regard to
+     * timestamps return by the used {@link org.apache.kafka.streams.processor.TimestampExtractor TimestampExtractor})
+     * and not based on wall-clock time.
      *
-     * @param timestamp the stream time when this method is being called
-     * @return new value; if null it will not be forwarded to down stream
+     * @param timestamp the stream time when {@code punctuate} is being called
+     * @return must return {@code null}&mdash;otherwise, and {@link org.apache.kafka.streams.errors.StreamsException
+     * exception} will be thrown
      */
-    R punctuate(final long timestamp);
+    VR punctuate(final long timestamp);
 
     /**
      * Close this processor and clean up any resources.
+     * <p>
+     * It is not possible to return any new output records within {@code punctuate}.
```

Comment:
No. Must be `It is not possible to return any new output records within {@code close}.`